receiver/statsdreceiver/transport/unix_server.go

```diff
@@ -0,0 +1,133 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package transport // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver/transport"
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/multierr"
+)
+
+type unixServer struct {
+	listener    net.Listener
+	reporter    Reporter
+	connections map[net.Conn]struct{}
+	wg          sync.WaitGroup
+	lock        *sync.Mutex
+	socket      string
+}
+
+var _ (Server) = (*unixServer)(nil)
+
+// NewUnixServer creates a transport.Server listening from a Unix Socket.
+func NewUnixServer(socket string) (Server, error) {
+	err := os.Remove(socket)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("removing socket failed: %w", err)
+	}
+
+	listener, err := net.Listen("unix", socket)
+	if err != nil {
+		return nil, err
+	}
+
+	u := unixServer{
+		listener:    listener,
+		connections: make(map[net.Conn]struct{}),
+		wg:          sync.WaitGroup{},
+		lock:        &sync.Mutex{},
+	}
+	return &u, nil
+}
+
+func (u *unixServer) ListenAndServe(nextConsumer consumer.Metrics, reporter Reporter, transferChan chan<- Metric) error {
+	if nextConsumer == nil || reporter == nil {
+		return errNilListenAndServeParameters
+	}
+
+	var returnErr error
+	u.reporter = reporter
+
+	for {
+		conn, err := u.listener.Accept()
+		if err != nil {
+			returnErr = fmt.Errorf("accepting new connection: %w", err)
+			break
+		}
+
+		u.wg.Add(1)
```

Comment:
perhaps a simpler way to manage this, instead of wg and locks could be to add here a select where a case manage reads (closing connections) and a case for the total shutdown.

Reply:
I think that locks , wg and map of connections could be simplified if used that pattern